src/hotspot/share/nmt/vmatree.cpp

```diff
@@ -197,3 +197,57 @@ VMATree::SummaryDiff VMATree::register_mapping(position A, position B, StateType
   }
   return diff;
 }
+
+#ifdef ASSERT
+void VMATree::print_on(outputStream* out) {
+  visit_in_order([&](TreapNode* current) {
+    out->print(SIZE_FORMAT " (%s) - %s - ", current->key(), NMTUtil::tag_to_name(current->val().out.mem_tag()),
+               statetype_to_string(current->val().out.type()));
```

Comment:
Could we consider here adding a utility function:

```
  IntervalState getState(TreapNode* node) {
    return node->val().out;
  }
```

to simplify from:

```
    out->print(SIZE_FORMAT " (%s) - %s - ", current->key(), NMTUtil::tag_to_name(current->val().out.mem_tag()),
                statetype_to_string(current->val().out.type()));
```

to

```
    out->print(SIZE_FORMAT " (%s) - %s - ", current->key(), NMTUtil::tag_to_name(getState(current).mem_tag()),
                statetype_to_string(getState(current).type()));
```

and all the other from:

`    stB.out = leqA_n->val().out;
`

to:

`    stB.out = getState(leqA_n);
`

Reply:
Sure, but I had to pick `in_state` and `out_state` as there are two states we're interested in.